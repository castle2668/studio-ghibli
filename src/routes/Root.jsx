import React, { useEffect } from "react";
import {
  Form,
  NavLink,
  Outlet,
  useLoaderData,
  useNavigation,
  useSubmit,
} from "react-router-dom";

import { getFilms } from "@/utils/films";

// Here are standard web objects: Request, URL, URLSearchParams.
export async function loader({ request }) {
  if (!request) return {};
  const url = new URL(request.url);
  const q = url.searchParams.get("q");
  const films = await getFilms(q);
  return { films, q };
}

const Root = () => {
  let films;
  let q;
  const loaderData = useLoaderData();
  if (loaderData?.films) films = loaderData.films;
  if (loaderData?.q) q = loaderData.q;
  const navigation = useNavigation();
  const submit = useSubmit();

  // The `navigation.location` will show up
  // when the app is navigating to a new URL and loading the data for it.
  // Therefore, the spinner will go away when there is no pending navigation anymore.
  const searching = navigation?.location
    && new URLSearchParams(navigation.location.search).has("q");

  const getNavLinkClass = ({ isActive, isPending }) => {
    if (isActive) {
      return "active";
    }
    if (isPending) {
      return "pending";
    }
    return "";
  };

  useEffect(() => {
    document.getElementById("q").value = q;
  }, []);

  return (
    <>
      <div id="sidebar">
        <h1>Studio Ghibli Films</h1>
        <div>
          <Form id="search-form" role="search">
            <input
              id="q"
              className={searching ? "loading" : ""}
              aria-label="Search films"
              placeholder="Search"
              type="search"
              name="q"
              // JSX "defaultValue" prop is the original "value" field of HTML <Input> element
              // JSX "value" prop is in sync with React state
              defaultValue={q}
              // the `currentTarget.form` is the input's parent form node
              // The submit function will serialize and submit any form you pass to it.
              onChange={(event) => {
                // submit(event.currentTarget.form);
                const isFirstSearch = q == null; // if this is the first search or not
                submit(event.currentTarget.form, {
                  replace: !isFirstSearch, // only want to replace search results
                });
              }}
            />
            <div id="search-spinner" aria-hidden hidden={!searching} />
            <div className="sr-only" aria-live="polite" />
          </Form>
        </div>
        <nav>
          {films && films.length ? (
            <ul>
              <li>
                <NavLink to="films" className={getNavLinkClass} end>
                  All Films
                </NavLink>
              </li>
              {films.map((film) => (
                <li key={film.id}>
                  <NavLink to={`films/${film.id}`} className={getNavLinkClass}>
                    {film.title}
                    {" "}
                    {film.favorite === "true" && <span>â˜…</span>}
                  </NavLink>
                </li>
              ))}
            </ul>
          ) : (
            <p>
              <i>No films</i>
            </p>
          )}
        </nav>
      </div>
      <div
        id="detail"
        className={navigation?.state === "loading" ? "loading" : ""}
      >
        <Outlet />
      </div>
    </>
  );
};

export default Root;
