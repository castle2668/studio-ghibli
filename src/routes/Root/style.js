import styled, { keyframes } from "styled-components";

const spinAnimation = keyframes`
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
`;

export const Sidebar = styled.div`
  width: 352px;
  background-color: #f7f7f7;
  border-right: solid 1px #e3e3e3;
  display: flex;
  flex-direction: column;

  & > * {
    padding-left: 32px;
    padding-right: 32px;
  }

  & h1 {
    font-size: 16px;
    font-weight: 500;
    display: flex;
    align-items: center;
    margin: 0;
    padding: 16px 32px;
    border-top: 1px solid #e3e3e3;
    order: 1;
    line-height: 1;
  }

  & h1::before {
    content: "";
    display: inline-block;
    background-image: url(@/assets/images/Studio_Ghibli_old.webp);
    background-size: contain;
    background-repeat: no-repeat;
    background-position: center center;
    width: 55px;
    height: 20px;
    margin-right: 8px;
    position: relative;
    top: 1px;
  }

  & > div {
    padding-top: 16px;
    padding-bottom: 16px;
    border-bottom: 1px solid #e3e3e3;
  }

  & > div form {
    position: relative;
  }

  & > div form input[type="search"] {
    width: 100%;
    padding-left: 32px;
    background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' class='h-6 w-6' fill='none' viewBox='0 0 24 24' stroke='%23999' stroke-width='2'%3E%3Cpath stroke-linecap='round' stroke-linejoin='round' d='M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z' /%3E%3C/svg%3E");
    background-repeat: no-repeat;
    background-position: 10px 12px;
    background-size: 16px;
    position: relative;
  }

  & > div form input[type="search"].loading {
    background-image: none;
  }

  /* NAV */
  & nav {
    flex: 1;
    overflow: auto;
    padding-top: 16px;
  }

  & nav a span {
    float: right;
    color: #eeb004;
  }
  & nav a.active span {
    color: inherit;
  }

  i {
    color: #818181;
  }
  & nav .active i {
    color: inherit;
  }

  & ul {
    padding: 0;
    margin: 0;
    list-style: none;
  }

  & li {
    margin: 4px 0;
  }

  & nav a {
    display: flex;
    align-items: center;
    justify-content: space-between;
    overflow: hidden;

    white-space: pre;
    padding: 8px;
    border-radius: 8px;
    color: inherit;
    text-decoration: none;
    gap: 16px;
  }

  & nav a:hover {
    background: #e3e3e3;
  }

  & nav a.active {
    background: hsl(224, 98%, 58%);
    color: white;
  }

  & nav a.pending {
    color: hsl(224, 98%, 58%);
  }
`;

export const SearchSpinner = styled.div`
  width: 16px;
  height: 16px;
  background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' fill='none' viewBox='0 0 24 24'%3E%3Cpath stroke='%23000' strokeLinecap='round' strokeLinejoin='round' strokeWidth='2' d='M20 4v5h-.582m0 0a8.001 8.001 0 00-15.356 2m15.356-2H15M4 20v-5h.581m0 0a8.003 8.003 0 0015.357-2M4.581 15H9' /%3E%3C/svg%3E");
  animation: ${spinAnimation} 1s infinite linear;
  position: absolute;
  left: 10px;
  top: 12px;
`;

export const Input = styled.input`
  font-size: 16px;
  font-family: inherit;
  border: none;
  border-radius: 8px;
  padding: 8px 12px;
  box-shadow: 0 0px 1px hsla(0, 0%, 0%, 0.2), 0 1px 2px hsla(0, 0%, 0%, 0.2);
  background-color: white;
  line-height: 1.5;
  margin: 0;

  &:hover {
    box-shadow: 0 0px 1px hsla(0, 0%, 0%, 0.6), 0 1px 2px hsla(0, 0%, 0%, 0.2);
  }
`;

export const SrOnly = styled.div`
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border-width: 0;
`;

export const Detail = styled.div`
  flex: 1;
  padding: 32px 64px;
  width: 100%;
  overflow-y: scroll;

  &.loading {
    opacity: 0.25;
    transition: opacity 200ms;
    transition-delay: 200ms;
  }
`;
