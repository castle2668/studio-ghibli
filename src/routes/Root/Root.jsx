import React, { useEffect } from "react";
import {
  Form,
  NavLink,
  Outlet,
  useLoaderData,
  useNavigation,
  useSubmit,
} from "react-router-dom";

import { getFilms } from "@/utils/films";

import * as Style from "./style";

// Here are standard web objects: Request, URL, URLSearchParams.
export async function loader({ request }) {
  const url = new URL(request.url);
  const q = url.searchParams.get("q");
  const films = await getFilms(q);
  return { films, q };
}

const Root = () => {
  const { films, q } = useLoaderData();
  const navigation = useNavigation();
  const submit = useSubmit();

  // The `navigation.location` will show up
  // when the app is navigating to a new URL and loading the data for it.
  // Therefore, the spinner will go away when there is no pending navigation anymore.
  const searching = navigation.location
    && new URLSearchParams(navigation.location.search).has("q");

  const getNavLinkClass = ({ isActive, isPending }) => {
    if (isActive) {
      return "active";
    }
    if (isPending) {
      return "pending";
    }
    return "";
  };

  useEffect(() => {
    document.getElementById("q").value = q;
  }, []);

  return (
    <>
      <Style.Sidebar>
        <h1>Studio Ghibli Films</h1>
        <div>
          <Form id="search-form" role="search">
            <Style.Input
              id="q"
              className={searching ? "loading" : ""}
              aria-label="Search films"
              placeholder="Search"
              type="search"
              name="q"
              // JSX "defaultValue" prop is the original "value" field of HTML <Input> element
              // JSX "value" prop is in sync with React state
              defaultValue={q}
              // the `currentTarget.form` is the input's parent form node
              // The submit function will serialize and submit any form you pass to it.
              onChange={(event) => {
                // submit(event.currentTarget.form);
                const isFirstSearch = q == null; // if this is the first search or not
                submit(event.currentTarget.form, {
                  replace: !isFirstSearch, // only want to replace search results
                });
              }}
            />
            <Style.SearchSpinner aria-hidden hidden={!searching} />
            <Style.SrOnly aria-live="polite" />
          </Form>
        </div>
        <nav>
          {films.length ? (
            <ul>
              <li>
                <NavLink to="films" className={getNavLinkClass} end>
                  All Films
                </NavLink>
              </li>
              {films.map((film) => (
                <li key={film.id}>
                  <NavLink to={`films/${film.id}`} className={getNavLinkClass}>
                    {film.title}
                    {" "}
                    {film.favorite === "true" && <span>â˜…</span>}
                  </NavLink>
                </li>
              ))}
            </ul>
          ) : (
            <p>
              <i>No films</i>
            </p>
          )}
        </nav>
      </Style.Sidebar>
      <Style.Detail className={navigation.state === "loading" ? "loading" : ""}>
        <Outlet />
      </Style.Detail>
    </>
  );
};

export default Root;
