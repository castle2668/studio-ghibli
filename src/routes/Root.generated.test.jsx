import React from "react";
import renderer from "react-test-renderer";

import Root, { loader } from "./Root";

jest.mock("react-router-dom");
jest.mock("@/utils/films");

const renderTree = (tree) => renderer.create(tree);
describe("<Root>", () => {
  it("should render component", () => {
    expect(renderTree(<Root />).toJSON()).toMatchSnapshot();
  });
});

describe("loader", () => {
  it("should expose a function", () => {
    expect(loader).toBeDefined();
  });

  it("loader should return expected output", async () => {
    const retValue = await loader({});
    expect(retValue).toEqual({});
  });
});
