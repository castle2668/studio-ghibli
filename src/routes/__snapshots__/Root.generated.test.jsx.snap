// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Root> should render component 1`] = `
[
  <div
    id="sidebar"
  >
    <h1>
      Studio Ghibli Films
    </h1>
    <div />
    <nav>
      <p>
        <i>
          No films
        </i>
      </p>
    </nav>
  </div>,
  <div
    className=""
    id="detail"
  />,
]
`;
